File: linked_list/linked_list.rb
class MyLinkedList
    attr_accessor :head, :size, :tail
=begin
    Initialize your data structure here.
=end
    def initialize()
      @head = nil
      @size = 0
      @tail = nil
    end
=begin
    Get the value of the index-th node in the linked list. If the index is invalid, return -1.
    :type index: Integer
   :rtype: Integer
=end
    def get(index)
      return -1 if @head == nil
      return -1 if index > @size - 1
      return @head.val if index == 0
      return @tail.val if index == @size - 1
      i = @head
      j = 0
      while i != nil
        if j == index
          return i.val
        end
        i = i.next
        j += 1
      end
      -1
    end
=begin
    Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list.
    :type val: Integer
    :rtype: Void
=end
    def add_at_head(val)
      new_node = Node.new(val)
      new_node.next = @head
      @tail = new_node if @head == nil
      @head = new_node
      @size += 1
    end
=begin
    Append a node of value val to the last element of the linked list.
    :type val: Integer
    :rtype: Void
=end
    def add_at_tail(val)
      return add_at_head(val) if @head == nil
      node = Node.new(val)
      @tail.next = node
      @tail = node
      @size += 1
    end
=begin
    Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted.
    :type index: Integer
    :type val: Integer
    :rtype: Void
=end
    def add_at_index(index, val)
      return add_at_head(val) if index == 0
      return add_at_tail(val) if @head == @tail
      return add_at_tail(val) if @size == index
      return if index > @size
      i = @head
      j = 0
      while i != nil
        if j == index - 1
          tmp = i.next
          node = Node.new(val)
          i.next = node
          node.next = tmp
          @size += 1
          return
        end
        i = i.next
        j += 1
      end
    end
=begin
    Delete the index-th node in the linked list, if the index is valid.
    :type index: Integer
    :rtype: Void
=end
    def delete_at_index(index)
      return if index < 0 || index > @size - 1
      return delete_at_head if index == 0
      return delete_at_tail if index == @size - 1
      i = @head
      j = 0
      while i != nil
        if j == index - 1
         i.next = i.next.next
         @size -= 1
         return
        end
        i = i.next
        j += 1
      end
    end

    def delete_at_head
      if @head == nil
        return nil
      elsif @head == @tail #one element in the linked list
        @head = nil
        @tail = nil
        @size -= 1
        return
      else
        @head = @head.next
        @size -= 1
      end
    end

    def delete_at_tail
      if @head == nil
        return nil
      elsif @head == @tail
        return delete_at_head
      end
      current = @head
      prev = nil
      while(current != tail)
        prev = current
        current = current.next
      end
      prev.next = nil
      @tail = prev
      @size -= 1
      return
    end

    private

    class Node
        attr_accessor :val, :next
        def initialize(val)
            @val = val
            @next = nil
        end
    end
end
# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList.new()
# param_1 = obj.get(index)
# obj.add_at_head(val)
# obj.add_at_tail(val)
# obj.add_at_index(index, val)
# obj.delete_at_index(index)
# Input
# ["MyLinkedList","addAtHead","get","addAtHead","addAtHead","deleteAtIndex","addAtHead","get","get","get","addAtHead","deleteAtIndex"]
#
# [[],[4],[1],[1],[5],[3],[7],[3],[3],[3],[1],[4]]
#
# Output
# [null,null,4,null,null,null,null,4,4,4,null,null]
# Expected
# [null,null,-1,null,null,null,null,4,4,4,null,null]
obj = MyLinkedList.new()
binding.pry
obj.add_at_head(0)
obj.get(1)
obj.add_at_index(1,4)
obj.add_at_tail(8)
obj.add_at_head(5)
obj.add_at_index(4,3)
obj.add_at_tail(0) 
obj.add_at_tail(5)
obj.add_at_index(6,3)
obj.delete_at_index(7)
obj.delete_at_index(5)
obj.add_at_tail(4)
binding.pry

File: linked_list/linked-list.js
var Node = function(val) {
  this.val = val;
  this.next = null;
}


/**
 * Initialize your data structure here.
 */
var MyLinkedList = function() {
  this.head = null;
  this.size = 0;
  this.tail = null;
};

/**
 * Get the value of the index-th node in the linked list. If the index is invalid, return -1. 
 * @param {number} index
 * @return {number}
 */
MyLinkedList.prototype.get = function(index) {
  if(this.head === null) { return -1 }
  if(index > this.size - 1) { return -1 }

  if(index === 0) {
    return this.head.val;
  }

  if(index === this.size - 1) {
    return this.tail.val;
  }

  let j = 0;
  for(let i = this.head; i !== null; i = i.next) {
    if(j === index){
      return i.val;
    }
    j++;
  }
};

/**
 * Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list. 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtHead = function(val) {
  let node = new Node(val);

  node.next = this.head;
  if(this.head === null) {
    this.tail = node;
  }
  this.head = node;
  this.size++;
  return
};

/**
 * Append a node of value val to the last element of the linked list. 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtTail = function(val) {
  if(this.head === null) {
    return this.addAtHead(val);
  }

  let node = new Node(val);
  this.tail.next = node;
  this.tail = node;
  this.size++;
};

/**
 * Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted. 
 * @param {number} index 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtIndex = function(index, val) {
  if(index === 0) {
    return this.addAtHead(val);
  }
  if(index === this.size) {
    return this.addAtTail(val);
  }
  if(index > this.size) { return }

  let j = 0
  for(let i = this.head; i !== null;i = i.next) {
    if(j === index - 1) {
      let temp = i.next
      let node = new Node(val);
      i.next = node;
      node.next = temp;
      this.size++;
      return
    }
    j++;
  }
};

/**
 * Delete the index-th node in the linked list, if the index is valid. 
 * @param {number} index
 * @return {void}
 */
MyLinkedList.prototype.deleteAtIndex = function(index) {
  if(index < 0 || index > this.size - 1) { return }
  if(index === 0) {
    return this.deleteAtHead();
  }

  if(index === this.size - 1) {
    return this.deleteAtTail();
  }
  let j = 0;
  for(let i = this.head; i !== null; i = i.next) {
    if(j === index - 1) {
      i.next = i.next.next;
      this.size--;
      return;
    }
    j++;
  }

};

MyLinkedList.prototype.deleteAtHead = function() {
  if(this.head === null) {
    return
  }

  if(this.head === this.tail){
    this.head = null;
    this.tail = null;
    this.size--;
    return
  }

  this.head = this.head.next;
  this.size--;
  return
};

MyLinkedList.prototype.deleteAtTail = function() {
  if(this.head === null){ return }
  if(this.head === this.tail){
    return this.deleteAtHead();
  }

  let current = this.head;
  let prev = null;

  while(current !== null) {
    prev = current
    current = current.next;
  }
};

/** 
 * Your MyLinkedList object will be instantiated and called as such:
 * var obj = new MyLinkedList()
 * var param_1 = obj.get(index)
 * obj.addAtHead(val)
 * obj.addAtTail(val)
 * obj.addAtIndex(index,val)
 * obj.deleteAtIndex(index)
 */

// https://leetcode.com/problems/design-linked-list/discuss/647240/JavaScript%3A-Faster-than-93-of-all-submissions
